### PR TITLE
Modern Event System: fix EnterLeave plugin logic

### PR DESCRIPTION
**Pull request by trueadm**
Today we found an internal issue with the modern event system and the EnterLeave plugin. Specifically, we didn't track `mouseout` on the container, only in the portal. That's because the portal had a child contained `onMouseEnter`. This bug is actually very similar to that of the recent fix to https://github.com/facebook/react/pull/18720.

Making this change, so that we now register this plugin on each root container fixed that issue, but also brought about another issue. This was great though, as we had already solved this in the legacy EnterLeave plugin (I commented on it below). Taking that logic and bringing it to the modern event system fixed the failing issues and aligned the modern event system with the legacy one.

### CODE REVIEW

File: packages/react-dom/src/client/ReactDOMRoot.js [hunk 1]
```diff
@@ -129,6 +132,11 @@ function createRootImpl(
         ? container
         : container.ownerDocument;
     eagerlyTrapReplayableEvents(container, doc);
+  } else if (enableModernEventSystem) {
+    // We use ensureListeningTo because the container might
+    // be a COMMENT_NODE and this function handles events
+    // on comments for us.
+    ensureListeningTo(container, 'onMouseEnter');
```
Comment (trueadm) [inline packages/react-dom/src/client/ReactDOMRoot.js:139]: This is similar to what we did with the fix in #18720.

File: packages/react-dom/src/events/plugins/ModernEnterLeaveEventPlugin.js [hunk 2]
```diff
@@ -165,6 +162,14 @@ const EnterLeaveEventPlugin = {
     enter.target = toNode;
     enter.relatedTarget = fromNode;
 
+    // If we are not processing the first ancestor, then we
+    // should not process the same nativeEvent again, as we
+    // will have already processed it in the first ancestor.
+    const nativeTargetInst = getClosestInstanceFromNode(nativeEventTarget);
```
Comment (trueadm) [inline packages/react-dom/src/events/plugins/ModernEnterLeaveEventPlugin.js:168]: This is basically the same as what we do on:

https://github.com/facebook/react/blob/master/packages/react-dom/src/events/plugins/LegacyEnterLeaveEventPlugin.js#L168-L173

File: packages/react-dom/src/client/ReactDOMComponent.js [hunk 3]
```diff
@@ -280,16 +279,15 @@ function ensureListeningTo(
...
+    if (
       rootContainerElement != null &&
-        rootContainerElement.nodeType === ELEMENT_NODE,
-      'ensureListeningTo(): received a container that was not an element node. ' +
-        'This is likely a bug in React.',
-    );
-    listenToEvent(registrationName, ((rootContainerElement: any): Element));
+      rootContainerElement.nodeType === ELEMENT_NODE
+    ) {
+      listenToEvent(registrationName, ((rootContainerElement: any): Element));
```
Comment (trueadm) [inline packages/react-dom/src/client/ReactDOMComponent.js:289]: We don't need an invariant here, we can use this more as a utility function. This let's us re-use this function and it's logic

File: packages/react-dom/src/client/ReactDOMRoot.js [hunk 4]
```diff
@@ -129,6 +132,10 @@ function createRootImpl(
         ? container
         : container.ownerDocument;
     eagerlyTrapReplayableEvents(container, doc);
+  } else if (enableModernEventSystem) {
+    // We use ensureListeningTo because the container might
+    // be a COMMENT_NODE.
```
Comment (gaearon) [inline packages/react-dom/src/client/ReactDOMRoot.js:137]: What does this have to do with comment nodes? I'm not following. The test case doesn't use them so I'm not seeing the connection.

File: packages/react-dom/src/client/ReactDOMComponent.js [hunk 5]
```diff
@@ -280,16 +279,15 @@ function ensureListeningTo(
       rootContainerInstance.nodeType === COMMENT_NODE
         ? rootContainerInstance.parentNode
         : rootContainerInstance;
-    // Containers can only ever be element nodes. We do not
+    // Containers should only ever be element nodes. We do not
     // want to register events to document fragments or documents
     // with the modern plugin event system.
-    invariant(
```
Comment (gaearon) [inline packages/react-dom/src/client/ReactDOMComponent.js:286]: Why relax the invariant?

Comment (trueadm) [reply packages/react-dom/src/client/ReactDOMComponent.js:286]: Yeah, I'll bring it back in.

File: packages/react-dom/src/client/ReactDOMRoot.js (hunk 4 above)
Comment (trueadm) [reply packages/react-dom/src/client/ReactDOMRoot.js]: I've update the condition above. The comment is redundant now.
